### ISSUE TITLE
Integration of new Bevy Book

### ISSUE THREAD

**Issue by Captain-Of-Coit** [bevyengine/bevy-website#623]
After talking with @alice-i-cecile for a bit on Discord, I ended up writing a small document (first just for myself) to highlight the current progress on the new book, and how we could move forward with it. Sharing it here to get it signed off on, so I could proceed to actually implement it.

**End goal**
Bootstrap new Official Bevy Book and enable documentation contributions.

**Approach**
Setup a basic structure for each section of the new book and add versions so Bevy Book 0.11.0 corresponds to Bevy Engine 0.11.0 Merge book to main (but hidden from public website) that can be iterated on until it feels complete enough to publish to the public, hopefully together with 0.11.0.

**Motivation**
New book is a bit stuck in limbo. Better documentation is in dire need, but contributions to it are currently missing, and hard to organize. A small focus on integrating the new book into `main` asap will make it easier to organize contributions and get it into a initial version to be published.

**Related RFCs**

- RFC#23 - https://github.com/bevyengine/rfcs/blob/main/rfcs/23-quick_start_book.md
	- Information + Structure for Bevy Book and new Quickstart

**Branches**

- bevyengine/bevy-website#new-book - https://github.com/bevyengine/bevy-website/compare/main...new-book - initial changes by @alice-i-cecile for new structure
- akappel/bevy-website#new-book-redux -> new-book - https://github.com/bevyengine/bevy-website/compare/main...akappel:bevy-website:new-book-redux - some additions + split of quick-start and book by @akappel
    - PR here: https://github.com/bevyengine/bevy-website/pull/602

**Missing**

- [ ] Hide book from public for now, only accessible by URL
- [ ] Versioned by Bevy version
- [ ] Basic scaffolded structure
- [ ] Disallow indexing robots for new Bevy book for now (added after @alice-i-cecile comment below)
- [ ] Add Book contribution guidelines outlining workflow described in this issue's discussion (added after @cart comments below)
- [ ] Merge to `main`

**Implementation**
- Hide book from public:
	- Rename "The Bevy Book" link to "Bevy Quickstart"
	- Change URL (with redirect)
		- From: https://bevyengine.org/learn/book/introduction/
		- To:      https://bevyengine.org/learn/quick-start
- Versioned by Bevy version
	- Book should be accessible URL:
		- https://bevyengine.org/learn/book/v0.11.0/introduction/
	- Each Bevy version gets its own version of the book by copy-paste
		- Note: We could do something fancy, like abstracting out the code samples and version specific things into their own files, include those in particular documentation files and somewhat resuse things when they don't change between versions. But feels needlessly complicated when just text content is not that expensive to copy-paste
		- Note: This would make the previous book versions somewhat "immutable" in that we don't change content one version before the latest one, it's only the current version that changes.
			- Example: Book 0.11 is released with Bevy 0.11. Bevy 0.12 was released with Book 0.12, but typo in Book 0.11 was found. Is it worth changes both versions? Could add a lot of maintaince burden.
	- Versions should be accessible by a dropdown menu that shows all available versions.
- Make sure all existing website links redirect to the correct location in the new information location, if changed
- Book structure from RFC #23 implementated (https://github.com/bevyengine/rfcs/blob/main/rfcs/23-quick_start_book.md) as scaffolding
- Once three previous steps are completed, we can merge the work into main and publish new website. The new Bevy book won't be visible (but accessible by URL) until book is polished enough. Contributions to `main` regarding the Bevy book can now be encourged as there is a minimal structure to work against.

--------------

@alice-i-cecile If I can get a acknowledgment that this is a OK way to proceed, I can take ownership over getting these changes through and hopefully continue to drive the effort of getting great documentation in place. This is obviously just the first stepping stone, but hopefully it can get the ball rolling

**Comment by alice-i-cecile** [comment]
I agree on the general direction. Getting this live sooner is better. Can we disable search engine indexing for the non-linked pages too?

> Example: Book 0.11 is released with Bevy 0.11. Bevy 0.12 was released with Book 0.12, but typo in Book 0.11 was found. Is it worth changes both versions? Could add a lot of maintaince burden.

No: this way lies madness. We don't backport bug fixes for Bevy itself, we shouldn't do this for the book.

> Each Bevy version gets its own version of the book by copy-paste
Note: We could do something fancy, like abstracting out the code samples and version specific things into their own files, include those in particular documentation files and somewhat resuse things when they don't change between versions. But feels needlessly complicated when just text content is not that expensive to copy-paste

Agreed: let's keep this simple.

I've read through the details of this plan and endorse it. Let's get this moving again: thanks for picking this up!

**Comment by Captain-Of-Coit** [comment]
> Can we disable search engine indexing for the non-linked pages too?

Yeah, sure thing. Shouldn't get indexed regardless as the various crawlers shouldn't find any of the entry-points, but better safe than sorry, so will make sure it's explicitly disallowed 👍 

For the rest: Thanks :)

**Comment by cart** [comment]
I also agree with this general direction! I do think the copy/paste versioning approach is probably the right path forward. It does create some challenges around widgets / styles etc. Ex: if we change the interface of a widget (ex: the zola shortcode changes, the styling changing, etc), that means we might break _all_ of the historical versions of the book. I don't love the idea of the copy/paste approach for widget code, styles, etc so we'd be committing to backporting that sort of thing indefinitely. In practice I think we'd need to ensure that every change we make to "the book writing api + style" is backportable in an automated way.

Detecting breakage / botched api backporting will also be challenging (and increasingly so as the number of versions increases).

It also creates asset-management issues. Assets (images, videos, sounds, etc) will often be "big" so we don't want to duplicate them an unbounded number of times. But it _also_ means that we need their paths to be stable indefinitely. We can't remove assets (or move them around) without breaking historical versions of the book. Therefore we should probably have a global "book assets" folder with strict rules for anything inside it (no deletes, renames are only allowed with a full find-and-replace ... and even then they should be discouraged because relative paths are challenging to fix up).

We should have answers to these problems before moving forward.

**Comment by alice-i-cecile** [comment]
I'm also okay with cosmetic (but not functional) breakage to old versions of the book: these are much less important and readers should have some sense of what they're getting into.

We could also build e.g. PDF versions of the old books and then only host those, taking down the other ones completely. This would solve the widget problem, but it would suck to break / degrade links.

**Comment by Captain-Of-Coit** [comment]
@cart thanks for the feedback, you raise good points. Sorry for the incoming blob of text, but it feels important we get all of this right from the get go before locking us into something.

Short summary of various versioning headaches:

- Written content
- Widgets
- Binary assets
- Error checking code samples

**Written content**

Written content can be solved with relative ease and the content (in terms of bytes) is small too. Copy-paste would do for now.

**Widgets**

Regarding widgets (Zola shortcodes), it can addressed with the following praxis (shamelessly stolen from how to deal with "no-change" legacy code but still wanting to refactor)

- "Never modify or remove arguments from function (shortcode) signatures, default to existing behaviour"
    - As far as I understand Zola, we can support this https://www.getzola.org/documentation/content/shortcodes/
- Pro: Existing widget remain functional
- Con: shortcodes grow in complexity, every additional parameter usage needs to be "guarded" in {% if variable %} { variable } { endif }, defaulting to the behavior and look as it was before introducing the parameter.

Overall, I don't know exactly what you expect in the future, but I'm unsure widgets will change drastically across versions so that the additional complexity becomes much of a concern. Going this route, we should have the increase in our mind though.

**Assets**

I think what you wrote would mostly work, except I'd argue that assets should be 100% immutable. If edits are made, they should be duplicated and then changed instead. So basically a "append-only" directory. Files made for a specific version can have a "v0.11.0" suffix as an example, or grouped by directories by version. If none of the images from 0.11.0 have changed for 0.12.0 documentation, all of them would refer to `/book/assets/0.11.0/`. Regardless, we should probably leverage git LFS (https://git-lfs.com/) for the assets directory as well, as it helps to manage binary blobs.

**Error checking code samples**

I think we can get away with that once a version documentation has been finalized, it shouldn't have to have its samples checked anymore, as it should be frozen in time together with which Bevy version was used. If there are major bugfixes that gets backported for a previous version, it'll be a new version of the book (imagine 0.10.2 happens after releasing 0.11.0 for example), and can be checked with that particular version in mind for the release, and not needed to be checked again afterwards.

-------- 

Pondering all of this though, I'm wondering if maybe we're approaching this wrong in general, or if there already are existing solutions to this. And the obvious answer is that yes, we're not the first ones to think about all of this :)

One obvious contender to consider is https://readthedocs.org/ which is widely used in FOSS, and provides most of the features we need for the documentation out of the box.

No copy-paste would be needed as versioning is supported out of the box, each version can simply be a git tag and it all gets built together. Widgets and assets become locked in time for their version as well, when leveraging the git history for what it does best, versioning. `main` branch would track `main` branch in Bevy, and each Bevy version gets its corresponding git tag when it's released.

With that approach, we get a different set of work to be done, but it solves a lot of other problems. Biggest one I can see is writing our own theme to fit with Bevy's existing theme on the website, and integrating the header from the Bevy website. Otherwise it seems like a relatively smooth way of proceeding, as an alternative.

**Comment by mockersf** [comment]
git lfs quotas on GitHub make it mostly unusable. There is a 1GB download monthly quota. The repo `content` folder is 134MB, so that's less than 10 downloads per month. And downloads in GitHub Actions count against the quota

**Comment by Captain-Of-Coit** [comment]
> Yup there are plenty of options out there. From my perspective owning our own stack is a win:

Yeah, certainly makes a lot of sense 👍 

> If this is really what we want, we could also easily do this. Separate repo for content, site book pages are generated on build by grabbing each bevy version tag. 

Considering we probably want to do the least amount of work to get the ball rolling in a somewhat proper way, I'd say we leave any further improvements out for now, and can be revisited in the future, if a strong need arise. For now, stricter processes would do the trick.

> git lfs quotas on GitHub make it mostly unusable. There is a 1GB download monthly quota. The repo content folder is 134MB, so that's less than 10 downloads per month. And downloads in GitHub Actions count against the quota

That certainly makes it unusable from the get go, only have experience using it with Gitea and Huggingface which are a bit more generous than that, thought GitHub was as well... Oh well 🤷 

-------------

@cart so it seems we're in agreement on what needs to be done, as far as I can tell. After our discussion I added another task as a requirement for merging the book: 

"Add Book contribution guidelines outlining workflow described in this issue's discussion (added after @cart comments below)"

In this, I'll start it out with a condensed and concise summary of what we've discussed in this issue, so we can hopefully steer contributions to adhere to it.

But I think at this point we nailed down what's needed in order to at least have a first, non-public version of the book that we could start accepting contributions to.

Once I receive your approval, I'll begin making it a reality :)

**Comment by cart** [comment]
Yup we're in agreement. You have my approval!

**Comment by Captain-Of-Coit** [comment]
PR to get this in place here: https://github.com/bevyengine/bevy-website/pull/624

**Comment by IceSentry** [comment]
I'm not a fan of versioning the book personally. At least, not that granularly. 

I think we should have a latest and main version, but until we reach 1.0 I really don't think we should keep outdated and unsupported documentation on the public internet. We can already see this issue all the time with docs.rs and google, where google will just point people to super old versions of a library.

If we had any kind of support guarantee for older versions, then sure, versioning could make sense but right now it seems like it adds complexity for very little benefits.

In other words, I think the versioning should reflect the development of bevy and the book is intended for newcomers to bevy. Experienced users on older versions will want docs for their version, but the book isn't targeted to them.

**Comment by alice-i-cecile** [comment]
Hmm. I would be on board with the proposal to cut versioning from the book, at least pre 1.0: I think it simplifies things considerably and I'm similarly concerned about links to outdated information.

**Comment by Captain-Of-Coit** [comment]
Some previous discussions about versioning of the documentation (besides the automatically generated documentation from code):

- https://github.com/bevyengine/bevy-website/issues/77
- https://github.com/bevyengine/bevy-website/issues/388

Summarized pros & cons as I understand it currently:

Cons:
- Introduces complexity right now by having multiple versions
- Search engines might link to old versions, confusing people

Pros:
- Users who use older version of Bevy than `main` or latest stable also gets proper documentation
- Links don't ever break
- Versioning is most likely gonna be needed post 1.0.0 regardless
- Decouples releasing documentation from deploying documentation (everyone can contribute to `main` in this repository)

> @IceSentry I think we should have a latest and main version

As I understand this, we'll still be having versions in this case, just that they become `whatever is the latest stable release` and `main`, which is the development version. So the complexity is already created in this case, and we'll have to manage search engines linking to potentially dead links (something was at `/learn/book/latest/ecs/stages` before, in newest `latest`, it doesn't exists so dead link).

And if we're doing those two versions, making a explicit `x.x.x` version is simply taking `latest` and copying it to the appropriate version directory. It's basically needed regardless if we chose to do `latest` / `main` split.

So realistically, we're not choosing between "having numbered releases available" vs "only two versions available", but the choice should be between "no versions" and "versions", as otherwise the two "con" arguments won't apply.

That leaves us with just supporting the currently stable Bevy version, as the alternative to "numbered versions".

And if we go that way, we're back to the square where things cannot get merged into `bevyengine/bevy-website` until the latest version has been merged, as then it'll track `main` rather than latest stable version.

Even when I try to see as many positives from not having any versioning at all for the Book, I don't see how we'll end up in a better situation than having versioning. Not only from a user perspective, from a "contributor to the website" perspective as well, things become easier to maintain, not harder. Similarly to using feature flags for decoupling releasing features from deploying features.

> @IceSentry If we had any kind of support guarantee for older versions

I didn't think there was any "support guarantee" at all, regardless of version, as is common with most FOSS. Helping is self-served, but we could assist people stuck on older version to get help easier, rather than having to pull down the bevy-website repo, find the right commit, check it out, install Zola, build the website.

**Comment by IceSentry** [comment]
To help understand that whole wall of text, my assumptions for the book are that it is intended for beginners to learn the basics and advanced users shouldn't really have to use the book. 

I think versioning of api documentation is very different than versioning of a book where the intention is mostly for new users to get started. I don't think anyone should be reading a getting started book for version 0.9 when 0.10 is released and 0.11 is in progress. The users that are stuck on old version of bevy are very limited and most likely experienced users that already have complicated project and wouldn't really derive value from the book anyway. For example, the rust book technically has multiple versions online but you can't really access any of them unless you actively try to do that because most people should be using the latest version. There's no dropdown on the site itself to find those versions and even then, they only have first-edition and second-edition.

By latest and main, I mostly meant that the main version of the book shouldn't be easily accessible until the related version of bevy is actually released. I think this is a pretty different flow than keeping a separate version for every bevy version. Honestly, I'd even be fine with only keeping a branch of the main version and not make it public or do something similar to dev-docs.bevyengine.org where we simply have a version of the website that tracks main and one that tracks latest. The public website would only have one version of the book. 

Also, adding content to the book could be made in the context of the latest version, especially with the current state of the book where it's definitely not complete. For things that can't be expressed for latest, then I don't really see an issue with keeping a branch for that. The new-book branch has been an issue because it wasn't updated frequently, but once the infrastructure is in place keeping a branch with only content changes based on main doesn't seem like a big issue. It would just be one more item in the release checklist to merge it. That's essentially how I've been doing the migration guide for the last 3 versions of bevy. I just create a PR with the new migration guide page and update it over a few weeks. I also suspect that as time goes on we'll see some release that essentially don't need to touch the book except maybe for some superficial things.

The only people using the main version should be book contributors/reviewers, it would be nice if it was easier for them to do so, of course, but it's not something I consider a priority since it's most of these people already have the repo cloned anyway.

**Comment by cart** [comment]
> And if we go that way, we're back to the square where things cannot get merged into bevyengine/bevy-website until the latest version has been merged, as then it'll track main rather than latest stable version.

I do agree that this specifically should change no matter what we choose. People should be able to easily preview and edit of the next version of the book. (ideally with working "improve this page" links) 

> If we had any kind of support guarantee for older versions, then sure, versioning could make sense but right now it seems like it adds complexity for very little benefits.

I do buy this argument. We will want this eventually but I don't think it is necessary (or particularly useful relative to the maintenance cost + complexity) right now.

I think I'm starting to have a bias toward "`dev` branch auto-deployed to a different github pages instance" (for now). It introduces complications because people need to target a different branch. But that does create a reasonably simple story when it comes to "what is dev and what is release". And it opens the doors to working on other parts of the website in a similar fashion (although in most cases we should probably still do other development on main).

**Comment by cart** [comment]
In practice this would probably look like:

* Create a new `bevy-website-dev` (or maybe `bevyengine-org-dev`) repo
* Whenever we push to the `dev` branch in `bevy-website`, push that to the main branch of `bevy-website-dev` (using a CI job)
* Configure `dev.bevyengine.org` to point to `bevy-website-dev` github pages deployment.
* Configure robots.txt to ignore crawling on `dev.bevyenigne.org`

I chose `dev.bevyengine.org` for parity with `dev-docs.bevyengine.org` (and the `dev` branch convention). But if we want to reserve that for some future use and/or make it scarier we could adopt the `unstable.bevyengine.org` and `unstable-docs.bevyengine.org` convention.

**Comment by alice-i-cecile** [comment]
Still neutral on versioning, but chiming in on a few relevant points.

> I disagree with that the book contents is only valuable for beginners, it's useful for both intermediate and advanced users to have a reference to look up things you forget.

I certainly intend to be targeting a pretty broad audience here once the current "getting started" content is replaced.

> Long kept alive branches is where work goes to die, short and quickly integrated branches is where work gets done.

Having experimented with long branches for the new-book, I'm reluctant to encourage them again. It's really hard for others to pitch in or get started with long-lived branches.

**Comment by djeedai** [comment]
I mostly agree with the consensus here. Couple of comments:

- Unity mixes book and API, and version the ensemble. So it's a bad example, because the versioning is imposed by the API changes; the book ("Manual") part seldomly changes between releases unless a feature drastically changed.
- The argument for not versioning a book (as opposed to API docs) is pretty strong today. However once Bevy has multiple stable versions supported (like Unity) this will need to happen. And that's a tough issue, for which separating the content (needs per-Bevy-version versioning) from the form (can and should keep style etc. consistent across book versions) is almost surely the best approach. From my past experience on another project, trying to mix the two is a major pain.

**Comment by Captain-Of-Coit** [comment]
> The argument for not versioning a book (as opposed to API docs) is pretty strong today

As I see it, no is arguing for not versioning the book at all. The disagreement comes from how to do the versioning. There will be at least two versions (`main` and `latest stable`), the question is how that is achieved.

- If we do one git branch that is `dev` which tracks Bevy `main`, then either someone has to write the infra for keeping that branch automatically up to date, or someone has to commit to making sure that branch is up to date. My own experience with trying to have branches for N months is that it's something you should avoid. I'm not ready to be that person, nor to maintain a solution which keeps the branches up to date. As far as I understand current maintainers, they're also strapped on time as they are committed to other efforts.
- If we do `main` in bevy-website tracks `main` in Bevy, then no separate branch has to be made, and no one has to either write the infra or keep the branch up to date manually. The version switch from `main` to `latest` happens with one action once every 3 months (following Bevy release schedule), so in that case, I can personally commit to maintaining that, if no other maintainer wants to take care of that.

I'm mainly interested in getting the ball rolling for getting proper docs in place for more people, as it's a complaint that seems to come up often enough. I'm not interested in committing myself to additional work and additional time commitment to maintain git branches. I am interested in driving this change across the finish line and I could commit to cutting the release once every 3 months however.

**Comment by djeedai** [comment]
Publishing the public book from `main` when we release every 3 months and publishing continuously to the hidden dev book, and never backporting anything to stable/public between releases (like we do for code), is not versioning. And I believe that's what's proposed here.

**Comment by IceSentry** [comment]
Sorry, it's a bit hard to keep things short 😅 

I'm not necessarily saying to go full long-lived branches every release. My opinion is that most of the work on the book shouldn't need to be tied to a specific version of bevy. I believe writing new chapters should be done in context of latest not main. Most bevy concepts haven't changed all that much for a while. Yes, there's been some new features and syntax changes a bit here and there, but the concepts are still largely the same. I'll definitely expect syntax changes in the code samples, but I don't think the book will need a new long lived branch for every version. I think those syntax changes could be handled easily close to the release, essentially, what is already done with the current book. Of course, things like stageless would require a big change to a few chapters of the book, but most changes aren't like that and don't force you to change your mental model of everything. (This assumes that the book focuses more on concepts, which I believe is the current intention)

If this above assumption is wrong and we end up needing major changes every time, then yes, more structured versioning will be necessary. The code validation in the new-book branch (might have been merged in main already, not sure about that) should also help with fixing the syntax changes fairly quickly.

My main objection originally was mostly with the keeping a version of the book for each version of bevy. I'm not necessarily against the idea of versioning if it makes the main/latest split easier to deal with. I just don't think keeping older versions around is worthwhile right now.

**Comment by Captain-Of-Coit** [comment]
As I understand @IceSentry, they're proposing the following workflow:

- bevyengine/bevy starts working on next version, this is on the `main` git branch in that repository
- bevyengine/bevy-website starts keeping the docs up to date, tracking that branch on a `dev` branch or similar, in this repository
- Any contribution during this period gets split into two, once set goes to the `dev` branch (what's changed in `main`), other goes to the `main` branch in this repository (general changes affected the website, or affecting latest released version) and gets back-ported to the `dev` branch
- Once bevyengine/bevy releases a new version, `dev` branch gets merged into `main`

Maybe I'm misunderstanding what they were proposing, but that's how I understood it.

**Comment by djeedai** [comment]
Yes but there's no versioning happening here 😊

Also I disagree about the split contribution. We're not making any code change to stable between releases, so I don't see why docs should be different. Using the same model for the book would help users understand the model. We simply release the book once every three months, same as the code.

**Comment by IceSentry** [comment]
Here's what I'm proposing in a more structured manner:

* bevy keeps doing what it does
* bevy-website also keeps everything on main as much as possible:
	* New chapters are made within the context of latest
	* Changes that can't be done in the context of latest will need a branch, but I don't think they will be common.
	* Changes to the syntax should be done close to the release and in a PR just for that. 
		* The code validation should make it fairly easy to identify
		* This is important to not do it too soon because some of those syntax changes on main don't even last until the final release, so we shouldn't update the book with them until we know they will actually ship.

Again, this is all based on the assumption that the book teaches mostly concepts that don't change that often, syntax and api is covered and versioned by docs.rs and examples.

**Comment by Captain-Of-Coit** [comment]
Thank you for the structured list, that's helpful :) 

So with this in mind, if we now want to get the ball rolling on getting the Book out in time for 0.11, how do you suggest we approach this without A) publishing the book to the public while B) continue enabling contributions to the general website and C) facilitate contributions to the book that is being worked on?

It sounds like we'd have to make one branch, have all the changes there, until it's 100% ready, every contribution to the book has to go to that branch instead of `main`, and any files this branch touches has to be disallowed to be changed in `main` until the branch been merged, in order to avoid conflicts. Maybe there is a different way I'm missing?

**Comment by IceSentry** [comment]
Here's how I personally see it:

* I'm fine with your approach of making the current book the quick-start guide and having the new book a separate thing. 
* Merge the new book as-is, but don't mention or link it anywhere
	* There should be a warning somewhere on that page if anyone ends up on the URL, but I think it's fine if it lives on main already. For example, this reminds me of the new react docs site. It was in beta for over a year (maybe more, not sure), it was also public, but with a clear indication that it was incomplete and to use at your own risk.

**Comment by djeedai** [comment]
Is the issue that the Bevy website publishes continually from `main` and so we can't do dev work on docs without changes being immediately public? If so I guess a `book-dev` branch is the simplest and most practical solution for now, yes.

**Comment by Captain-Of-Coit** [comment]
> Here's how I personally see it:

Got it, thanks!

So let's say we hide it until it's ready, once it's ready we start linking it. All is well, 0.11.x book is now live and all is dandy. Work on 0.12 starts, what's the process? The book is already viewable so any changes into `main` will be visible, is this the case when we start doing a git branch instead, keep it open for ~3 months until next version of Bevy been released?

> Is the issue that the Bevy website publishes continually from main and so we can't do dev work on docs without changes being immediately public?

Yes

> If so I guess a book-dev branch is the simplest and most practical solution for now, yes.

Can we take that as you are volunteering for managing that branch while it remains open for ~3 months, making sure it's up to date with bevyengine/bevy-website#main ? 😄

**Comment by IceSentry** [comment]
> Work on 0.12 starts, what's the process? The book is already viewable so any changes into main will be visible, is this the case when we start doing a git branch instead, keep it open for ~3 months until next version of Bevy been released?

I don't think the book should try to follow main. Api level docs are covered by docs.rs and are updated alongside main already. The book should focus more on concepts, so if a concept doesn't have a chapter already, the new chapter should be done in the context of the latest release. 

For example, there's no chapter on rendering right now. If I want to write one, I should make sure to make it based on the latest released api, in your example that would be 0.11. If it's good enough to be merged, but not completely ready, then merge it and just don't link to it in the sidebar or anywhere else. Now that chapter will contain things like "to render something, you need to spawn an entity with those components" and then explain the purpose of those components. This concept hasn't change since pretty much bevy 0.1.0 so it doesn't need to be made with the context of 0.12.

Making changes to the book by adding new content can be done even if 0.11 is released as long as those changes are based on 0.11. If someone wants to explain in a bit more details how pipelined rendering works, they don't need bevy main to do that.

**Comment by IceSentry** [comment]
And if new chapters are made specifically for a new feature that only exists on main, it could also exists in an hidden page but still on main

**Comment by djeedai** [comment]
> Can we take that as you are volunteering for managing that branch while it remains open for ~3 months, making sure it's up to date with bevyengine/bevy-website#main ? 

`book-dev` stays alive forever alongside `main`, and merges into `main` on release. It contains only book changes; there's nothing to keep in sync.

> The book should focus more on concepts, so if a concept doesn't have a chapter already, the new chapter should be done in the context of the latest release not main.

The content of the book can be based on stable code or whatever, but the actual work of producing this content, like writing new code, doesn't _need_ to be published instantly; it can follow a 3-month cycle like the code itself. And so, every change to the book is published publicly on release day alongside new code. This limits churn and release overhead, same as code. It feels like we're trying to reinvent a solution for having instant book hotfix changes when we're not even doing that for the code, why?

**Comment by IceSentry** [comment]
> doesn't _need_ to be published instantly

It doesn't _need_ to, but in my opinion it should be. To me writing new chapters is in no way related to which bevy version is released other than syntax related changes and keeping this behind a branch that only gets merged on new releases doesn't really seem to add any value to me. If a chapter is ready to be published or if an existing chapter has been improved, why wait until a new version of bevy is released to publish it? Keeping a separate branch seems like more overhead than just publishing things when they are ready and only updating the syntax when close to release if necessary.

For example, the learn-wgpu book releases chapters when they are ready and gets syntax update when a new version of wgpu is released, but it doesn't wait until a new version of wgpu to release a chapter if it's ready.

**Comment by Captain-Of-Coit** [comment]
> The book should focus more on concepts

> This concept hasn't changed since pretty much bevy 0.1.0 so it doesn't need to be made with the context of 0.12.

Yes, this is true, but the book also contains a lot of examples, that uses the syntax of a specific version. This is outlined in the RFC (https://github.com/bevyengine/rfcs/blob/main/rfcs/23-quick_start_book.md), which is linked in the very opening of this issue. The RFC contains exactly what the book should contain, so there isn't much point in trying to redefine that here.

While concepts might not change drastically between versions, there is a lot of stuff that does change between versions, and a lot of those things will be used throughout the book. 

As a sneak-peek, you can browse the [new-book's branch content](https://github.com/bevyengine/bevy-website/tree/new-book/content/learn/book), and see that once completed, there will be multiple changes across versions.

> so if a concept doesn't have a chapter already

> just don't link to it in the sidebar or anywhere else

Each addition to the book will probably be mentioned at multiple points of the book, linking to a central location. Just hiding that specific page will in many cases not be enough. It's not simply "add a new page" and book is done, stuff is gonna have to be rewritten and so on.

> book-dev stays alive forever alongside main, and merges into main on release. It contains only book changes; there's nothing to keep in sync.

There will be changes that goes into main, that will have to be merged into book-dev as well. This is upkeep. 

> It doesn't need to, but in my opinion it should be

I agree that it probably should, but I don't think we should block the solution of this particular topic that this issue is about in order to have that. It's a different discussion and a different issue.

-----------

With all of this said, I think I've spent well enough of my own and others time regarding this, so I think at this point I'm at a inflection point where I'll defer to @cart / @alice-i-cecile exactly on how to proceed, as there doesn't seem to be any immediate consensus. 

I also have a slight feeling that there is a lot of context and information in this issue's conversation and linked resources in the initial issue body that seemingly have been glossed over, which feels a bit like we're loosing time discussing the wrong things, or re-iterating on previously mentioned things. It's possible I haven't been clear enough as well, and if so, I apologize for that.

What the Book is has been defined already, the process and whatever this issue resolves as, need to support whatever is defined in there. If you want to change the meaning of the book, I think the right process is the RFCs, which is where it was initially defined and agreed upon.

It might be useful for everyone to also state outright if they're willing to block what's the proposed (as-is) in the issue's list named "Missing". We're not writing anything in stone, I'm just trying to get the ball rolling. If we do it one way and it's wrong, we can always adjust.

**Comment by IceSentry** [comment]
I'm aware of the book definition in the RFC and everything I said is specifically based on that definition. What I'm trying to say is that syntax changes between versions are largely trivial especially in short and focused examples which is exactly how the book is described in the RFC and how the wip book already is. 

I don't see the point of adding versioning/separate branches just to handle changes that I personally consider trivial to do once per release. It just seems like we disagree on the complexity of those changes between version. The book RFC was merged 2 years ago and every subject it wants to cover is just as relevant today and the vast majority of it hasn't seen major syntax changes. This is precisely why I don't want to add complexity in the form of a separate branch and/or versioning.

This doesn't apply to the quick start game section of the RFC, but I assumed this was only about the book and quick start games are a separate issue entirely. Versioning for those would be valuable.

To circle back to the missing section of this issue. The only thing I don't like is the versioning, everything else I'm 100% in favor, especially the merge to main but keep it hidden part.

**Comment by B-Reif** [comment]
> * Unity mixes book and API, and version the ensemble. So it's a bad example, because the versioning is imposed by the API changes; the book ("Manual") part seldomly changes between releases unless a feature drastically changed.

Unity’s conflation of book content and API documentation is awful to deal with. I agree that it’s not a good comparison point.

Are we doctesting the book? If so, that will reveal any breaking syntax changes. We can reasonably expect larger changes—the kind where we would expect an RFC—to be integrated into the book as soon as they’re available.

The API documentation provides granular content for both reference and discovery. I think it follows that the book’s role is focused on broader concepts that change more slowly. Anything granular enough to ‘not work’ should be caught by doc tests.

For these reasons, I’m comfortable with a looser coupling of book and Bevy version.

**Comment by IceSentry** [comment]
> Are we doctesting the book?

I don't remember if it's just on the new-book branch or main, but yes, the code is validated

**Comment by B-Reif** [comment]
- After 1.0, breaking changes will be denoted by major version numbers. Semver gives us a clean break between compatible versions of the book.
- Before 1.0, we don't currently track breaking changes on a per-version basis. Most `0.x` versions ship breaking changes, and they're listed in the patch notes, but not strictly tracked.
- The book is intended to cover topics from a broader perspective than API documentation. On a per-version basis, each release already has its own API documentation.

For all of these reasons, I'm against a separate branch or versioning while Bevy is still on v0. Once we're adhering to semver, we'll know when to fork the book. Before then, we're already explicitly making no guarantees about breaking changes per-version.

**Comment by djeedai** [comment]
> We can reasonably expect larger changes—the kind where we would expect an RFC—to be integrated into the book as soon as they’re available.

That means we'll have the official Bevy book referencing some concepts and code that is not supported / doesn't exist in the latest stable version. This is probably going to confuse users.

**Comment by IceSentry** [comment]
It seems to me that "as soon as they're available" implies that it would only be publicly accessible once the new api is released.

**Comment by cart** [comment]
First: thank you all for the thoughtful discussion. A lot of good ideas have been discussed here and I think it is pretty clear that most of the approaches discussed would work well. We are also in the fortunate situation where switching between approaches is pretty trivial / superficial, so we don't _need_ to make this perfect right out of the gate. The second we reach consensus on pain points or a better path forward, we can always make changes.

In the interest of moving forward I'm making a call, but this is _just to get the ball rolling_ (which is the most important thing by a pretty wide margin). This conversation doesn't have to be over.

This is what we should do for now:

### 1. The `bevy-website/main` branch contains `bevy/latest` content.

The primary process for contributing content changes is "create a pull request targeting `bevy-website/main`":
* If a user reading the book clicks "improve this page", it links to editing the file on the main branch
* We should limit the amount of "unstable Bevy Book content authoring" because it is a moving target. More on this later.
   
### 2. Sections that are under construction or aren't _quite_ ready are merged in the "draft" state

* Draft content should not be directly discoverable by users and crawlers, but links should still work
* This will use the `draft` variable in zola in combination with a tool to scan for draft pages and generate `robot.txt` block lists (we've built similar front matter tools in the past ... this should not be hard).
* We will render "draft" pages with some prominent "under construction" warning at the top and maybe some other style changes (could use some variant of the http://dev-docs.bevyengine.org logo).
* The `draft` state  _is not a dumping ground for unreviewed content_:
    * The content _can_ be incomplete, but it should be designed with the public eye in mind. We should be reasonably confident in the quality of lines of text that _do_ exist. If something is missing, that should be indicated. We should adopt a standardized way of denoting a `TODO` section (maybe with special styling via some `todo` shortcode).
    * The structure/organization doesn't need to be perfect for drafts, but it should be intentional (and that structure / organization should be reviewed in PRs). I think people will naturally try to help "fill in" missing sections, so we should treat section placeholders responsibly / as a part of the planning process.
* This is _not_ the "one size fits all" solution for all in-progress content. But it should cover many common cases reasonably well:
    * If we are adding new sections that aren't quite ready yet, we can merge progress iteratively with the draft flag for easier collaboration ... notably this should work well for bootstrapping the Bevy Book. But also for bootstrapping new sections going forward.
    * If we are making sweeping changes to existing non-draft sections, they should be scrutinized appropriately in the PR and merged when ready for public consumption
* Some exceptional cases:
    * Large / sweeping organizational reworks with massive cross-section reference/linking implications
        * these should be discouraged generally, broken up into smaller pieces whenever possible, and completed as quickly as possible to avoid merge-conflict-churn.
        * when they absolutely must happen, they should be done on a central coordinating branch with an owner responsible for facilitating the effort and keeping the branch up to date
        * If a group of sections are _replacing_ other sections, we can allow them to exist in parallel (with the new versions in draft mode). This should help prevent long-running branches / merge conflict resolution. Once the new sections are ready to be swapped in, we can do a final "fix-up links / other indirect references and un-draft" PR.
* It would be cool if "draft" sections could link to a `bevy-website` issue (specified in the section front-matter) that helps interested users track the progress on the section (and find ways to contribute if they are interested).

### 3. Authoring "unstable / next release" Bevy Book content _will_ still need to happen sometimes

* Some examples:
    * Adding a new section to the book to cover a new feature not covered in `latest`
    * Updating to use new APIs / syntax
* In these cases ... we will use a separate `release-0.11` branch:
    * It provides a central point for coordination, when it is needed (ex: for extended collaborative documentation efforts).
    * It forces us to contend with merge conflicts with the `bevy/latest` (`bevy-website/main`) version of the book ... this is a good thing! Work being done on `latest` sometimes won't line up perfectly with work being done for the next release. If we just use side-by-side version folders, then identifying and resolving conflicts actually gets harder + riskier. Git and merge conflict identification + resolution tooling are our friends here.
    * We have the option to host this under a different URL, but I think we should hold off on this until there is demand (and only after we've gone through the process at least once without it).
* I suspect that managing this branch should generally not be too much effort, but rest assured that for now, maintainers (especially me as the one full time / fully paid maintainer) will be on the hook for managing it (if nobody else volunteers). I'm not forcing anyone to facilitate this process.

### 4. We don't _have_ a `bevy/latest` book yet! What process do we use?
 
All content will be added in `draft` mode to `bevy-website/main` and the entire top level book will be hidden. We can choose to take content out of `draft` mode on a case by case basis whenever it is ready, likely with an intentional "first" release with a chunk of consistent content that can make a splash.

### Next Steps

* The discussion on this issue has gotten long / hard to follow, and it has become more of a meta "how should we do this" issue. We should keep it scoped to that. Feel free to respond to this with additional ideas and thoughts. When we are ready to "kick off" the effort, we should close this and open a new issue. @Captain-Of-Colt if you are still interested in driving this effort, create a new issue to capture this new process and track progress. I'm also happy to do that.
* Implement "draft" mode style / rendering
* _We need to sort out how "draft section" development will work_. I think we will want a way to opt-in to rendering draft sections in the sidebar for easy discoverability during the development process. If zola's built-in `draft` field won't work for this, we should evaluate other options.
  * This is most important for developers. Ex: being able to define some global "unhide drafts" option when generating the html from templates. _However_ there is a world where we would want this to be dynamic (ex: use javascript to show/hide draft sections in the sidebar using url config: `https://bevyengine.org/learn/book/introduction?show_draft=true`). I believe this would be possible without too much effort (ex: generate javascript _inside_ the template that feeds draft page status into global some javascript state, then show/hide content in the sidebar using that status).
* Implement "draft" mode robots.txt generation
* Define and implement "TODO section"  style / wording / shortcode (as in `h2`, `h3` section)
* Implement the top-level hiding of the Bevy Book

On this comment, please react with:
* :+1:  this comment if you are _ok with moving forward with this approach for now_
* :-1:  (and explain in a comment) if you are against moving forward with this approach for now. This is not about what you think would be _best_. Please only respond with :-1: if you would prefer to block progress until I change some aspect of the process.

**Comment by cart** [comment]
_edited in one small point of clarity in section (1)_

**Comment by cart** [comment]
Looks like we have unanimous support for moving forward. @Captain-Of-Coit has volunteered to drive the effort. @Captain-Of-Coit: create a new issue to track progress / summarize the approach (feel free to copy paste anything from above) and we'll go from there!

**Comment by alice-i-cecile** [comment]
Tracking in #870 :)
